### PR TITLE
Add CuTe W4A16 inference dot kernels

### PR DESCRIPTION
**Pull request by ry2009**
## Summary

This PR directly addresses the original kernel request: take the forward passes in `compressed/mlx.py` and `compressed/awq.py`, add custom GPU fast paths, and make them work under `vmap` so both scalar matvec decode and batched matmul/prefill route correctly.

H100 80GB HBM3, `6144 x 2048`, `group_size=32` unless noted.

| original ask | before | after | speedup | route |
| --- | ---: | ---: | ---: | --- |
| `compressed/mlx.py` inference forward, matvec/decode | 301-303us | 6.79us | ~45x | JAX-native CuTe W4A16 |
| `compressed/awq.py` inference forward, matvec/decode | 301-303us | 6.76us | ~45x | JAX-native CuTe W4A16 |
| MLX inference forward under `vmap`, batch 4 matmul | 300-301us | 8.57us | ~35x | batched packed route |
| AWQ inference forward under `vmap`, batch 4 matmul | 300-301us | 8.54us | ~35x | batched packed route |
| MLX/AWQ inference forward under `vmap`, batch 8 matmul | ~300us | 12.58-12.61us | ~24x | batched packed route |
| large-prefill route once dense cache exists | ~300us materialize path | 9.8-10.1us | ~30x | cached dense XLA tensor-core dot |
| MLX deterministic training forward | 40.23us | 23.28us | 1.73x | grouped JAX/VJP |
| AWQ deterministic training forward | 42.41us | 24.87us | 1.71x | grouped JAX/VJP |
| MLX deterministic training backward | 68.00us | 54.81us | 1.24x | grouped JAX/VJP |
| AWQ deterministic training backward | 69.21us | 59.65us | 1.16x | grouped JAX/VJP |
| 16k eval, limit 300 | 3068.53s | 2114.33s | 1.45x | full model eval |

## Original Target Benchmark

H100 80GB HBM3, shape `6144 x 2048`, `group_size=32`, fp16/bf16 vectors. Baseline is the old materialize-dequant-plus-matmul path.

| batch | MLX packed int4 | AWQ packed int4 | old materialize+matmul | speedup |
| ---: | ---: | ---: | ---: | ---: |
| 1 | 6.79us | 6.76us | 301-303us | ~45x |
| 2 | 6.95us | 6.92us | 301-303us | ~44x |
| 4 | 8.57us | 8.54us | 300-301us | ~35x |
| 8 | 12.61us | 12.58us | ~300us | ~24x |
| 16 | 18.83us | 17.68us | ~300us | ~16-17x |
| 32 | 31.54us | 31.53us | ~300us | ~9.5x |
| 64 | 59.69us | 59.66us | 300-301us | ~5x |

Dense cached/pre-dequantized bf16 is still the large-prefill route for this shape: about `9.8-10.1us` once the dense weight exists. That is why the PR does not blindly call the packed GEMV for every batch size.

## Eval Result

Fast 16k MMLU-Pro Mirai eval, `--limit 300`.

| path | wall time | exact match | speedup |
| --- | ---: | ---: | ---: |
| baseline `/home/ry2009/lqd_pre_fused` | 3068.53s | 0.3874 +/- 0.0074 | 1.00x |
| fast path `/home/ry2009/lqd_pre_fused_densefast` | 2114.33s | 0.3855 +/- 0.0073 | 1.45x |

The eval speedup is smaller than the kernel speedup because the eval includes attention, KV/cache work, scheduling, sampling, and non-dot layers. Accuracy is within eval noise.

## Training Dot Result

Training stays in JAX. The grouped deterministic path avoids full dequant materialization for deterministic training matvec; backward is improved but still limited by the dense `dW` write.

| path | pass | original | grouped JAX/VJP | speedup |
| --- | --- | ---: | ---: | ---: |
| MLX | forward | 40.23us | 23.28us | 1.73x |
| MLX | backward | 68.00us | 54.81us | 1.24x |
| AWQ | forward | 42.41us | 24.87us | 1.71x |
| AWQ | backward | 69.21us | 59.65us | 1.16x |

Gradient relative RMS drift vs grouped JAX semantics: MLX `6.6e-4`, AWQ `1.1e-3`.

## Dispatch Rule

The mergeable JAX route is a cost model, not always custom kernel. Packed wins when packed memory traffic beats launch/custom-call overhead; dense wins when tensor cores or the cached dense weight dominate.

| case | route | reason |
| --- | --- | --- |
| decode, rows >= 10240 | CuTe packed GEMV | packed traffic wins once the output projection is large enough |
| decode, rows < 10240 | cached dense XLA dot | avoids custom-call overhead at smaller row counts |
| prefill, batch 2 and rows <= 4096 | CuTe packed GEMV | the only batched prefill case with a clean packed win |
| all other batched prefill | cached dense XLA dot | tensor-core dense dot beats GEMV-style packed kernels |
| explicit non-default `DotAlgorithmPreset` | existing JAX dot path | preserves the caller precision contract |

Recent route probes checked rows `2048/4096/6144/8192/10240/12288` and fp16/bf16. The final route keeps the simple, repeatable wins and avoids format/dtype-specific overfitting.

## CuTe vs TileLang

| comparison | result |
| --- | --- |
| TileLang raw fused GEMV best during earlier tuning | about 6.3us kernel-only |
| CuTe raw decode kernel profile during earlier tuning | about 5.8-5.9us kernel-only |
| mergeable JAX-native route | no Torch/TileLang bridge, respects `DotAlgorithmPreset`, works under external `vmap` |

TileLang was useful for proving the speed target. CuTe is what belongs in this PR because it is JAX-native and keeps the codebase/profile story clean.

## What Changed

- Added JAX-native CuTe/CUTLASS W4A16 inference dot kernels for MLX and AWQ output-input 4-bit group-32 matrices.
- Dispatches automatically from compressed MLX/AWQ inference dot when the supported fast-path contract holds.
- Supports fp16/bf16 outputs and external `vmap`.
- Keeps a dense dequantized inference cache rebuilt from packed parameters for tensor-core routes; exports still save only packed weights/scales/zero-points.
- Keeps deterministic training grouped-dot fast paths in JAX.
- Removed slower experimental dequant-to-global + GEMM work from this PR so the review surface stays small.

## Correctness

| check | result |
| --- | --- |
| MLX/AWQ packed routes vs materialized fp16 semantics | checked for batches 1, 2, 4, 8, 16 during tuning |
| route lowering checks | `6144` decode lowers to dense, `10240+` decode lowers to CuTe, `4096` batch-2 prefill lowers to CuTe, `2048` batch-4 prefill lowers to dense |
| dense inference cache export/load/sharding coverage | covered by compressed tests |
| H100 focused MLX/AWQ tests after latest route changes | `64 passed` |
| H100 compressed suite from earlier pass | `236 passed` |

## Checks Run

| command | result |
| --- | --- |
| `uv run --no-sync ruff check lalamo/compressed/mlx.py lalamo/compressed/awq.py lalamo/compressed/cute_w4a16_dot.py benchmarks/bench_w4a16_prefill.py benchmarks/__init__.py` | passed |
| `uv run --no-sync python -m py_compile lalamo/compressed/cute_w4a16_dot.py benchmarks/bench_w4a16_prefill.py` | passed |
| H100 `pytest tests/unit/compressed/test_mlx.py tests/unit/compressed/test_awq.py -q` | `64 passed` |
| H100 route probes over rows `2048/4096/6144/8192/10240/12288`, fp16/bf16 | completed |

## Follow-Up

Packed-MMA/tensor-core GEMM for replacing the dense cache in batch-prefill is split to `gup/packed-mma-prefill` / #239 / #238. Current stock CUTLASS mixed-input kernels do not solve exact MLX/AWQ group-32, so that work stays out of this merge PR.

### CODE REVIEW

**Comment by ry2009** [comment]
Superseded by the clean real-base branch in #240. This PR was stacked on `gup/gpu-kernels-brr-review`, so I am closing it to avoid merging through the closed GPU-kernel review stack.
